examples/tags.rs

```diff
@@ -38,36 +38,57 @@ fn main() {
 }
 
 fn run(host: &str, user: Option<String>, passwd: Option<String>, image: &str) -> Result<()> {
-    let mut tcore = try!(Core::new());
-    let mut dclient = try!(
-        dkregistry::v2::Client::configure(&tcore.handle())
-            .registry(host)
-            .insecure_registry(false)
-            .username(user)
-            .password(passwd)
-            .build()
-    );
+    let mut tcore = Core::new()?;
+    let client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(host)
+        .insecure_registry(false)
+        .username(user)
+        .password(passwd)
+        .build()?;
 
-    let futcheck = try!(dclient.is_v2_supported());
-    let supported = try!(tcore.run(futcheck));
-    if !supported {
-        return Err("API v2 not supported".into());
-    }
-
-    let fut_token = try!(dclient.login(&[&format!("repository:{}:pull", image)]));
-    let token_auth = try!(tcore.run(fut_token));
+    let futures = futures::future::ok::<_, dkregistry::errors::Error>(client)
+        .and_then(|dclient| {
+            dclient.is_v2_supported().and_then(|v2_supported| {
+                if !v2_supported {
+                    Err("API v2 not supported".into())
+                } else {
+                    Ok(dclient)
+                }
+            })
+        }).and_then(|dclient| {
+            dclient.is_auth(None).and_then(|is_auth| {
+                if is_auth {
+                    Err("no login performed, but already authenticated".into())
+                } else {
+                    Ok(dclient)
+                }
+            })
+        }).and_then(|dclient| {
+            dclient
+                .login(&[&format!("repository:{}:pull", image)])
+                .and_then(|token| {
+                    dclient
+                        .is_auth(Some(token.token()))
+                        .and_then(move |is_auth| {
+                            if !is_auth {
+                                Err("login failed".into())
+                            } else {
+                                println!("logged in!");
+                                Ok(dclient
+                                    .clone()
```

Comment:
Is this clone really needed?

Reply:
Yes, because:

```
error[E0597]: `dclient` does not live long enough
  --> examples/tags.rs:70:21
   |
69 |                 .and_then(|token| {
   |                           ------- capture occurs here
70 |                     dclient.is_auth(Some(token.token())).and_then(|is_auth| {
   |                     ^^^^^^^ borrowed value does not live long enough
...
82 |         }).and_then(|tags_collect| tags_collect);
   |         - borrowed value only lives until here
...
91 | }
   | - borrowed value needs to live until here

error[E0597]: `token` does not live long enough
  --> examples/tags.rs:76:49
   |
70 |                     dclient.is_auth(Some(token.token())).and_then(|is_auth| {
   |                                                                   --------- capture occurs here
...
76 |                                 .set_token(Some(token.token()))
   |                                                 ^^^^^ borrowed value does not live long enough
...
81 |                 })
   |                 - borrowed value only lives until here
...
91 | }
   | - borrowed value needs to live until here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0597`.
error: Could not compile `dkregistry`.
```